service/storage_service.cc

```diff
@@ -2077,8 +2078,47 @@ class topology_coordinator {
         slogger.info("raft topology: coordinator fiber found a node to work on id={} state={}", node.id, node.rs->state);
 
         switch (node.rs->state) {
-            case node_state::normal:
             case node_state::none: {
+                if (_topo_sm._topology.normal_nodes.empty()) {
+                    slogger.info("raft topology: skipping join node handshake for the first node in the cluster");
+                } else {
+                    auto validation_result = validate_joining_node(node);
+                    if (auto* reject = std::get_if<join_node_response_params::rejected>(&validation_result)) {
+                        // Transition to left
+                        topology_mutation_builder builder(node.guard.write_timestamp());
+                        builder.with_node(node.id)
+                               .del("topology_request")
+                               .set("node_state", node_state::left);
+                        auto reason = ::format("bootstrap: node rejected");
+                        co_await update_topology_state(std::move(node.guard), {builder.build()}, reason);
+
+                        slogger.info("raft topology: rejected node moved to left state {}", node.id);
+
+                        // Keep trying to send the rejection to the node, give up after some time.
+                        const auto send_reject_deadline = lowres_clock::now() + std::chrono::seconds(30);
+                        while (true) {
```

Comment:
Can this be done outside of a state machine? The state machine can proceed doing other topology operations while we try to kill a node. Also if coordinator changes after the node is marked left we will not try to shutdown. May be we should move to `node_state::left_token_ring` and handle the shutdown there they way we handle it during decommission: try one, if failed print a message that the node can be shutdown manually.

Reply:
It makes sense, but can we do it in a follow-up?

Reply:
We can.